ethcore/src/snapshot/service.rs

```diff
@@ -690,14 +717,20 @@ impl Service {
 	/// Feed a chunk of either kind (block or state). no-op if no restoration or status is wrong.
 	fn feed_chunk(&self, hash: H256, chunk: &[u8], is_state: bool) {
 		// TODO: be able to process block chunks and state chunks at same time?
-		let mut restoration = self.restoration.lock();
-		match self.feed_chunk_with_restoration(&mut restoration, hash, chunk, is_state) {
+		let r = {
+			let mut restoration = self.restoration.lock();
+			self.feed_chunk_with_restoration(&mut restoration, hash, chunk, is_state)
+		};
+		match r {
 			Ok(()) |
 			Err(Error::Snapshot(SnapshotError::RestorationAborted)) => (),
 			Err(e) => {
+				// TODO: after this we're sometimes deadlocked
 				warn!("Encountered error during snapshot restoration: {}", e);
-				*self.restoration.lock() = None;
-				*self.status.lock() = RestorationStatus::Failed;
+				self.abort_restore();
+				if let Some(mut status) = self.status.try_lock_for(std::time::Duration::from_millis(10)) {
```

Comment:
I'm not sure if `self.status.lock` can lead to a deadlock unless we acquire locks in different order. Could you elaborate on rationale for introducing `try_lock_for`?

Reply:
Rationale: this is non-crucial so if it can't take the lock that's ok. `self.status` is updated in several places and I did not want to spend too much time making sure they were ok.